packages/komencikit/src/kit.ts

```diff
@@ -0,0 +1,378 @@
+    )
+  }
+
+  /**
+   * submitMetaTransaction: uses the /v1/submitMetaTransaction endpoint
+   * It receives a wallet address and transaction (as a CeloTransactionObject)
+   * and creates a signature and passes everything to Komenci for execution
+   *
+   * @param walletAddress - the MTW that should execute the transaction
+   * @param tx - the transaction to be executed
+   * @param nonce - optional nonce to be used for signing the meta-tx
+   */
+  @retry({
+    tries: 3,
+    bailOnErrorTypes: [
+      FetchErrorTypes.Unauthorised,
+      FetchErrorTypes.ServiceUnavailable,
+      TxErrorTypes.Revert,
+    ],
+    onRetry: (_args, error, attempt) => {
+      console.debug(`${TAG}/submitMetaTransaction attempt#${attempt} error: `, error)
+    },
+  })
+  public async submitMetaTransaction(
+    walletAddress: string,
+    tx: CeloTransactionObject<any>,
+    nonce?: number
+  ): Promise<Result<TransactionReceipt, FetchError | TxError>> {
+    const wallet = await this.getWallet(walletAddress)
+    const signature = await wallet.signMetaTransaction(tx.txo, nonce)
+    const rawMetaTx = toRawTransaction(wallet.executeMetaTransaction(tx.txo, signature).txo)
+
+    const resp = await this.client.exec(submitMetaTransaction(rawMetaTx))
+    if (resp.ok) {
+      const txHash = resp.result.txHash
+      return this.waitForReceipt(txHash)
+    }
+
+    return resp
+  }
+
+  /**
+   * Utility function used to wait for a transaction to finalise and return the receipt
+   *
+   * @param txHash - the hash of the transaction to watch
+   * @private
+   */
+  private async waitForReceipt(txHash: string): Promise<Result<TransactionReceipt, TxError>> {
+    let receipt: TransactionReceipt | null = null
+    let waited = 0
+    while (receipt == null && waited < this.options.txRetryTimeoutMs) {
+      receipt = await this.contractKit.web3.eth.getTransactionReceipt(txHash)
+      if (receipt == null) {
+        await sleep(100)
+        waited += 100
+      }
+    }
+
+    if (receipt == null) {
+      return Err(new TxTimeoutError())
+    }
+
+    return Ok(receipt)
+  }
+
+  /**
+   * Utility function used to instantiate a MetaTransactionWalletWrapper cached by address
+   *
+   * @param txHash - the hash of the transaction to watch
+   * @private
+   */
+  _wallet?: MetaTransactionWalletWrapper
+  private async getWallet(address: string): Promise<MetaTransactionWalletWrapper> {
+    if (this._wallet === undefined || this._wallet.address !== address) {
+      this._wallet = await this.contractKit.contracts.getMetaTransactionWallet(address)
+    }
+    return this._wallet
+  }
+
+  /**
+   * Used to create a signature for a login message that allows the server to verify that the
+   * externalAccount passed in is actually owned by the caller.
+   *
+   * TODO: In order to increase security and prevent replay attacks we should introduce a nonce
+   * Flow would be:
+   * 1. KomenciKit asks for a nonce from Komenci
```

Comment:
Some ideas for a replay attack prevention:
1. Sign the `captchaResponseToken` since this can only be used once
2. Have the client generate the unique sessionId and sign this

Reply:
Ah the `catpcahResponseToken` option is really clever love it ❤️ .